geomesa-memory/src/main/scala/org/locationtech/geomesa/memory/cqengine/GeoCQEngine.scala

```diff
@@ -0,0 +1,188 @@
+* All rights reserved. This program and the accompanying materials
+* are made available under the terms of the Apache License, Version 2.0
+* which accompanies this distribution and is available at
+* http://www.opensource.org/licenses/apache2.0.php.
+*************************************************************************/
+
+package org.locationtech.geomesa.memory.cqengine
+
+import java.util
+import java.util.UUID
+
+import com.googlecode.cqengine.attribute.Attribute
+import com.googlecode.cqengine.index.hash.HashIndex
+import com.googlecode.cqengine.index.navigable.NavigableIndex
+import com.googlecode.cqengine.index.radix.RadixTreeIndex
+import com.googlecode.cqengine.index.unique.UniqueIndex
+import com.googlecode.cqengine.{ConcurrentIndexedCollection, IndexedCollection}
+import com.googlecode.cqengine.query.option.DeduplicationStrategy
+import com.googlecode.cqengine.query.{QueryFactory, Query}
+import com.googlecode.cqengine.query.simple.{Equal, All}
+import com.typesafe.scalalogging.LazyLogging
+import com.vividsolutions.jts.geom.Geometry
+import org.locationtech.geomesa.memory.cqengine.index.GeoIndex
+import org.locationtech.geomesa.memory.cqengine.utils._
+import org.locationtech.geomesa.utils.geotools._
+import org.opengis.feature.`type`.AttributeDescriptor
+import org.opengis.feature.simple.{SimpleFeature, SimpleFeatureType}
+import org.opengis.filter._
+
+import scala.collection.JavaConversions._
+
+class GeoCQEngine(sft: SimpleFeatureType,
+                  enableFidIndex: Boolean = false) extends LazyLogging {
+  //val cqcache = CQIndexingOptions.buildIndexedCollection(sft)
+  val cqcache: IndexedCollection[SimpleFeature] = new ConcurrentIndexedCollection[SimpleFeature]()
+  val attributes = SFTAttributes(sft)
+
+  // Add Geometry index on default geometry first.
+  // TODO: Add logic to allow for the geo-index to be disabled?  (Low priority)
+  addGeoIndex(sft.getGeometryDescriptor)
+
+  if (enableFidIndex) addFidIndex()
+
+  // Add other indexes
+  sft.getAttributeDescriptors.foreach {addIndex(_)}
+
+  def remove(sf: SimpleFeature): Boolean = {
+    cqcache.remove(sf)
+  }
+
+  def add(sf: SimpleFeature): Boolean = {
+    cqcache.add(sf)
+  }
+
+  def addAll(sfs: util.Collection[SimpleFeature]): Boolean = {
+    cqcache.addAll(sfs)
+  }
+
+  def clear(): Unit = {
+    cqcache.clear()
+  }
+
+  def getById(id: String): Option[SimpleFeature] = {
+    // if this gets used, set enableFidIndex=true
+    cqcache.retrieve(new Equal(SFTAttributes.fidAttribute, id)).headOption
+  }
+
+  def update(sf: SimpleFeature): Boolean = {
+    getById(sf.getID) match {
+      case Some(oldsf) => cqcache.remove(oldsf)
+      case None =>
+    }
+    cqcache.add(sf)
+  }
+
+  // NB: We expect that FID filters have been handled previously
+  def getReaderForFilter(filter: Filter): FR =
+    filter match {
+      case f: IncludeFilter => include(f)
+      case f => queryCQ(f, dedup = true)
```

Comment:
[picky] Seems strange to pass in "dedup = true" as it's already a default arg. Remove this param?